### DISCUSSION TITLE
How can I render an XFA form?

### DISCUSSION THREAD

**Discussion by hselvarajan** [pypdfium2-team/pypdfium2#221] [Q&A] [answered]
I'm trying to render an XFA form, for example the one here from iText:

https://github.com/itext/i7js-examples/blob/develop/src/main/resources/pdfs/purchase_order.pdf

With the default installation of pypdfium2, the XFA form is not rendered and the image below is rendered:

<img width="763" alt="image" src="https://github.com/pypdfium2-team/pypdfium2/assets/2226521/8eda169c-0b65-472d-873c-0aa226ad2746">

Does the pdfium library that is shipped with pypdfium2 include XFA (and therefore V8) support? If not, is there a way I can use my own library, like the one from here, which is built with XFA and V8 support?

https://github.com/bblanchon/pdfium-binaries/releases/download/chromium%2F5786/pdfium-v8-mac-arm64.tgz

Thank you for any help you can provide!

**Comment by mara004** [comment, marked as answer]
> Does the pdfium library that is shipped with pypdfium2 include XFA (and therefore V8) support?

Well, the wheels are packaged with the default (non-V8) pdfium builds. But if you install pypdfium2 from source, you can set an env var to use the V8 binaries (as mentioned [here](https://github.com/pypdfium2-team/pypdfium2#setup-magic)):
```bash
export PDFIUM_USE_V8=1
pip3 install git+https://github.com/pypdfium2-team/pypdfium2@stable
# alternatively: # pip3 install pypdfium2 --no-binary=pypdfium2
```

Then calling `pypdfium2 -v` would yield something like
```
pypdfium2 4.12.0 (libpdfium 5786, pdfium-binaries V8 build)
```

**Reply by mara004** [reply]
BTW, also note that, if you want to render forms, you must call `.init_forms()` on after `PdfDocument` construction, but probably you know this already.

**Comment by hselvarajan** [comment]
@mara004 Thank you very kindly for your advice. I installed pypdfium2 with the commands you suggested:

`export PDFIUM_USE_V8=1`
`pip3 install git+https://github.com/pypdfium2-team/pypdfium2@stable`

Any XFA form still fails to render, which is very strange. Some time ago I built a version of pdfium from source (off Chromium 4128) and its `pdfium_test` test program is able to render these files fine, so it's either an issue with bblanchon's pre-built libraries or a step that I'm missing in initializing the form environment. 

It also seems that, in order to work with XFA forms, one must invoke `FPDF_LoadXFA(doc)` after initializing the form environment, which I am doing as follows:

`print("Result from pdfium loadXFA: ", pdfium_c.FPDF_LoadXFA(pdf))`

`FPDF_LoadXFA` returns `0`, which is a failure code -- I wonder if this has anything to do with V8 initialization.

I'll try re-building pypdfium2 with pdfium from source, and see if that helps. Thank you!

**Reply by mara004** [reply]
> Any XFA form still fails to render, which is very strange. [...] so it's either an issue with bblanchon's pre-built libraries or a step that I'm missing in initializing the form environment.

Hmm, that's strange indeed. Just to confirm, you do call `.init_forms()` as I wrote above, right?

**Reply by mara004** [reply]
> It also seems that, in order to work with XFA forms, one must invoke FPDF_LoadXFA(doc) after initializing the form environment, which I am doing as follows:

Oh, thanks for the update - I guess we should also call FPDF_LoadXFA in `.init_forms()`.
I'll take a look.

**Reply by hselvarajan** [reply]
I do, indeed - here is the slightly hacky test program I'm using:

```python
import pypdfium2 as pdfium
import pypdfium2.raw as pdfium_c
pdf = pdfium.PdfDocument("/Users/hari/Downloads/email_recommended.pdf")
config = pdfium_c.FPDF_FORMFILLINFO(version=2, xfa_disabled = 0)
jsplatform = pdfium_c.IPDF_JSPLATFORM()
jsplatform.version = 3
config.jsPlatform = jsplatform
version = pdf.get_version()  # get the PDF standard version
print ("PDF Version: ", version)
pdf.init_forms()
print("Result from pdfium loadXFA: ", pdfium_c.FPDF_LoadXFA(pdf))
print ("PDF Form Type: ", pdf.get_formtype())

pagecount = len(pdf)

for i in range(pagecount):
    pdfpage = pdf.get_page(i)
    bitmap = pdfpage.render()
    pilimage = bitmap.to_pil()
    pilimage.show()
```

**Reply by mara004** [reply]
In your code, it looks like `config` isn't actually used - you still have to pass it to `init_forms()`, don't you?

**Reply by hselvarajan** [reply]
You're absolutely right, I was playing with using and not using the config, so I'd removed it before I pasted-in the code, but I just tried again while passing `config` to `init_forms()` and sadly it doesn't make a difference.

**Reply by mara004** [reply]
BTW, could you also share the other test file? I don't currently have a works-in-chromium XFA test file locally...

**Reply by hselvarajan** [reply]
[email_recommended.pdf](https://github.com/pypdfium2-team/pypdfium2/files/11574110/email_recommended.pdf)
Here you go. It's from the pdfium test suite:

https://pdfium.googlesource.com/pdfium/+/refs/heads/chromium/3062/test/email_recommended.pdf
(edit by mara004: file moved to) https://pdfium.googlesource.com/pdfium/+/refs/heads/main/testing/resources/xfa/email_recommended.pdf

**Reply by mara004** [reply]
Thanks. I tested and am making the same observation.

Until more or less recently, form initialization for pdfium-binaries V8 builds was broken, see https://github.com/bblanchon/pdfium-binaries/issues/105.
Maybe there could be a further issue with XFA?

**Reply by mara004** [reply]
I have now asked at pdfium-binaries: https://github.com/bblanchon/pdfium-binaries/issues/105#issuecomment-1565335503
If you have anything to add, please do so.

**Reply by mara004** [reply]
Hi @hselvarajan,

I think we finally found the issue, there was a typo in your code snippet that we both overlooked. It's `m_pJsPlatform`, not `jsPlatform`. Check out the ~~main~~ devel_v5 branch and call `init_forms()` with an XFA build.

**Reply by mara004** [reply]
However, while `email_recommended.pdf` now works, it looks like `purchase_order.pdf` is just not supported by pdfium.
It also does not render in Chromium despite `chrome://flags/#pdf-xfa-forms` set to `Enabled`.

**Reply by hselvarajan** [reply]
@mara004 Thank you so much for following this through to a resolution.  Sorry about the typo in the test program. I was on vacation which is why I couldn't respond sooner.

**Reply by mara004** [reply]
No problem, thanks for getting back to this! Hope you had a nice vacation.
ctypesgen should really error on typos in struct fields, see https://github.com/ctypesgen/ctypesgen/issues/183#issuecomment-1677258322. I hope to get that fixed too sooner or later.

**Reply by mara004** [reply]
FYI, the fix has been backported from the devel_v5 branch to main (f66c08ff9ab82a613ceac7a62389e759c1b0c12c) and will be included in the next release.

**Reply by mara004** [reply]
The ctypesgen issue is now fixed as well!
